### DISCUSSION TITLE
Suggestion: run *on* WASI

### DISCUSSION THREAD

**Discussion by gkgoat1** [libuv/libuv#4719] [Ideas]
Projects like `uvwasi` use `libuv` to provide to runtimes, and particularly compilers, WASI support. By having `libuv` support WASI (even if some system calls are stubbed out), further sandboxing and compilation would be available on `libuv`-based software, especially those which use `uvwasi`.

**Comment by bnoordhuis** [comment]
My hunch is that WASI is too anemic to target but paging @cjihrig, our resident expert.

**Comment by cjihrig** [comment]
I can't see a ton of usefulness for a platform abstraction library for wasm (since that is one of the selling points of wasm itself).

If I understand the request correctly, I don't think it would really help uvwasi either. First, uvwasi is a dead project. Some people may say otherwise, but look at the git history. uvwasi itself would also need to support running as wasm, and then embedders of uvwasi would need to migrate to the wasm build.

All of that said - if someone wanted to do all of the work to get libuv running on wasm, I wouldn't block it, but it doesn't seem worth it without a better use case.

I think **_if_** libuv wanted to be more useful to uvwasi, it would be via something like https://github.com/libuv/libuv/issues/4167 and https://github.com/libuv/libuv/pull/4434. But that wouldn't solve the dead project issue.

**Comment by gkgoat1** [comment]
> I can't see a ton of usefulness for a platform abstraction library for wasm (since that is one of the selling points of wasm itself).

I'm not interested in portability within WASI as much as portability between WASI and other targets.

> If I understand the request correctly, I don't think it would really help uvwasi either. First, uvwasi is a dead project. Some people may say otherwise, but look at the git history. 

It also is the most feature-complete implementation of WASI in C that I know of. It being dead may present challenges, especially if v2 comes before WASI becomes supported.

> uvwasi itself would also need to support running as wasm, and then embedders of uvwasi would need to migrate to the wasm build.

With `libuv` as a platform abstraction layer, `uvwasi` shouldn't have to support running as wasm. Embedders always could stay compiling to native: you'll even need to do that to run the code natively compiled. But the ability to create re-sandboxable code is important to embedders of a resultant binary (wasm2c, w2c2, any C runtime).

At least in v1, presenting WASI as a very odd UNIX would work, and special support for WASI could be a breaking change in v2.

> All of that said - if someone wanted to do all of the work to get libuv running on wasm, I wouldn't block it, but it doesn't seem worth it without a better use case.
> 
> I think **_if_** libuv wanted to be more useful to uvwasi, it would be via something like #4167 and #4434. But that wouldn't solve the dead project issue.

I want `libuv` to be more useful to WASI in general, not just `uvwasi`.

**Comment by cjihrig** [comment]
> I want libuv to be more useful to WASI

I think PRs are welcome as long as they aren't too intrusive.

**Comment by plajjan** [comment]
As a language implementer, I too am interested in libuv supporting wasm / wasi. So +1 to this work.

Libuv forms a central run time system component for the Acton programming language. Beyond the obvious I/I abstractions, libuv is also the core loop for the worker threads that runs the basic compiled language continuations. So even if we want to compile to wasm without io, we depend on libuv. I would really prefer to not have multiple run time loops but a single one that is using libuv. Acton is actor based and runs continuations concurrently across multiple worker threads. Libuv is used to wake up threads etc.

To contrast Libxev seems to offer better wasi support and thus the sort of cross platform support we are looking for. I think this style is the future.
